src/main/java/org/opensearch/neuralsearch/sparse/NeuralSparseIndexShard.java

```diff
@@ -0,0 +1,268 @@
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SegmentReader;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.SegmentInfo;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.util.BytesRef;
+import org.opensearch.common.lucene.Lucene;
+import org.opensearch.index.engine.Engine;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.index.shard.IndexShard;
+import org.opensearch.neuralsearch.sparse.codec.InMemorySparseVectorForwardIndex;
+import org.opensearch.neuralsearch.sparse.mapper.SparseTokensFieldType;
+import org.opensearch.neuralsearch.sparse.codec.InMemoryClusteredPosting;
+import org.opensearch.neuralsearch.sparse.common.InMemoryKey;
+import org.opensearch.neuralsearch.sparse.common.SparseVector;
+import org.opensearch.neuralsearch.sparse.codec.CacheGatedForwardIndexReader;
+import org.opensearch.neuralsearch.sparse.codec.CacheGatedPostingsReader;
+import org.opensearch.neuralsearch.sparse.codec.SparseTermsLuceneReader;
+import org.opensearch.neuralsearch.sparse.codec.SparseBinaryDocValuesPassThrough;
+import org.opensearch.neuralsearch.sparse.common.SparseVectorReader;
+import org.apache.lucene.index.SegmentReadState;
+import java.util.Set;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * NeuralSparseIndexShard wraps IndexShard and adds methods to perform neural-sparse related operations against the shard
+ */
+@Log4j2
+public class NeuralSparseIndexShard {
+    @Getter
+    private final IndexShard indexShard;
+
+    /**
+     * Constructor to generate NeuralSparseIndexShard. We do not perform validation that the index the shard is from
+     * is in fact a neural sparse Index (index.neural_sparse = true). This may make sense to add later, but for now the operations for
+     * NeuralSparseIndexShard that are not from a neural-sparse index should be no-ops.
+     *
+     * @param indexShard IndexShard to be wrapped.
+     */
+    public NeuralSparseIndexShard(IndexShard indexShard) {
+        this.indexShard = indexShard;
+    }
+
+    /**
+     * Return the name of the shards index
+     *
+     * @return Name of shard's index
+     */
+    public String getIndexName() {
+        return indexShard.shardId().getIndexName();
+    }
+
+    /**
+     * Load all the neural-sparse segments for this shard into the cache.
+     * Preloads sparse field data to improve query performance.
+     */
+    public void warmUp() throws IOException {
+        log.info("[Neural Sparse] Warming up index: [{}]", getIndexName());
+
+        final MapperService mapperService = indexShard.mapperService();
+
+        try (Engine.Searcher searcher = indexShard.acquireSearcher("neural-sparse-warmup")) {
+            for (final LeafReaderContext leafReaderContext : searcher.getIndexReader().leaves()) {
```

Comment:
Remember we have rank features segments, if it's rank feature segments, warm up and clear cache do nothing.